File: app/src/main/java/de/bitb/astroskop/ui/main/profiles/details/ProfileDetailFragment.java
package de.bitb.astroskop.ui.main.profiles.details;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import de.bitb.astroskop.AstroApplication;
import de.bitb.astroskop.R;
import de.bitb.astroskop.helper.Logger;
import de.bitb.astroskop.injection.IBind;
import de.bitb.astroskop.model.Constellation;
import de.bitb.astroskop.model.Profile;
import de.bitb.astroskop.ui.base.ActionbarHandler;
import de.bitb.astroskop.ui.base.BaseFragment;
import de.bitb.astroskop.ui.base.MVPFragment;
import de.bitb.astroskop.ui.main.profiles.details.add.CreateConstellationActivity;

import static android.app.Activity.RESULT_OK;
import static de.bitb.astroskop.ui.main.profiles.details.ProfileActivity.KEY_PROFILE_UUID;
import static de.bitb.astroskop.ui.main.profiles.details.add.CreateConstellationActivity.REQUEST_CREATE_CONSTELLATION;

public class ProfileDetailFragment extends MVPFragment<IProfileDetailView, ProfileDetailPresenter> implements IProfileDetailView, IBind {

    @BindView(R.id.fragment_profile_detail_aries_list)
    protected ViewGroup ariesList;
    @BindView(R.id.fragment_profile_detail_taurus_list)
    protected ViewGroup taurusList;
    @BindView(R.id.fragment_profile_detail_gemini_list)
    protected ViewGroup geminiList;
    @BindView(R.id.fragment_profile_detail_cancer_list)
    protected ViewGroup cancerList;
    @BindView(R.id.fragment_profile_detail_leo_list)
    protected ViewGroup leoList;
    @BindView(R.id.fragment_profile_detail_virgo_list)
    protected ViewGroup virgoList;
    @BindView(R.id.fragment_profile_detail_libra_list)
    protected ViewGroup libraList;
    @BindView(R.id.fragment_profile_detail_scorpio_list)
    protected ViewGroup scorpioList;
    @BindView(R.id.fragment_profile_detail_sagittarius_list)
    protected ViewGroup sagittariusList;
    @BindView(R.id.fragment_profile_detail_capricorn_list)
    protected ViewGroup capricornList;
    @BindView(R.id.fragment_profile_detail_aquarius_list)
    protected ViewGroup aquariusList;
    @BindView(R.id.fragment_profile_detail_pisces_list)
    protected ViewGroup piscesList;

    public static ProfileDetailFragment createInstance(String uuid) {
        ProfileDetailFragment frag = new ProfileDetailFragment();
        Bundle args = new Bundle();
        args.putString(KEY_PROFILE_UUID, uuid);
        frag.setArguments(args);
        return frag;
    }

    @Override
    protected ProfileDetailPresenter createPresenter(AstroApplication application) {
        return new ProfileDetailPresenter(application, this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getPresenter().onCreate(getArguments().getString(KEY_PROFILE_UUID));
    }

    @Override
    public ActionbarHandler.ActionbarCallback getActionbarCallback() {
        return new ActionbarHandler.ActionbarCallback() {
            @Override
            public int getActionbarButton1Icon() {
                return android.R.drawable.ic_menu_add;
            }

            @Override
            public boolean onActionbarButton1Clicked() {
                CreateConstellationActivity.startActivity(getActivity());
                return true;
            }
        };
    }

    @Override
    public void initProfile(Profile profile) {
        List<Constellation> constellations = profile.getConstellations();
        ariesList.removeAllViews();
        taurusList.removeAllViews();
        geminiList.removeAllViews();
        cancerList.removeAllViews();
        leoList.removeAllViews();
        virgoList.removeAllViews();
        libraList.removeAllViews();
        scorpioList.removeAllViews();
        sagittariusList.removeAllViews();
        capricornList.removeAllViews();
        aquariusList.removeAllViews();
        piscesList.removeAllViews();
        for (Constellation constellation : constellations) {
            ViewGroup viewGroup = null;
            switch (constellation.getZodiac()) {
                case ARIES:
                    viewGroup = ariesList;
                    break;
                case TAURUS:
                    viewGroup = taurusList;
                    break;
                case GEMINI:
                    viewGroup = geminiList;
                    break;
                case CANCER:
                    viewGroup = cancerList;
                    break;
                case LEO:
                    viewGroup = leoList;
                    break;
                case VIRGO:
                    viewGroup = virgoList;
                    break;
                case LIBRA:
                    viewGroup = libraList;
                    break;
                case SCORPIO:
                    viewGroup = scorpioList;
                    break;
                case SAGITTARIUS:
                    viewGroup = sagittariusList;
                    break;
                case CAPRICORN:
                    viewGroup = capricornList;
                    break;
                case AQUARIUS:
                    viewGroup = aquariusList;
                    break;
                case PISCES:
                    viewGroup = piscesList;
                    break;
            }
            boolean isHouse = constellation.isHouse();
            View view = LayoutInflater.from(getContext()).inflate(R.layout.profile_detail_constellation_list_item, null);
            TextView tv = view.findViewById(R.id.profile_detail_list_item_label);
            tv.setText(isHouse ? "House:" : "Ruler: ");
            tv = view.findViewById(R.id.profile_detail_list_item_text);
            tv.setText(isHouse ? constellation.getHouse().name() : constellation.getRuler().name());

            viewGroup.addView(view);
        }
    }

    @Override
    public void refreshView() {

    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_profile_detail;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK && requestCode == REQUEST_CREATE_CONSTELLATION) {
            getPresenter().addConstellation((Constellation) data.getSerializableExtra(CreateConstellationActivity.KEY_CONSTELLATION));
        }
    }
}

File: app/src/main/java/de/bitb/astroskop/ui/main/profiles/details/ProfileDetailFragment.kt
package de.bitb.astroskop.ui.main.profiles.details

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import butterknife.BindView
import de.bitb.astroskop.BaseApplication
import de.bitb.astroskop.R
import de.bitb.astroskop.injection.IBind
import de.bitb.astroskop.model.Constellation
import de.bitb.astroskop.model.Profile
import de.bitb.astroskop.ui.base.ActionbarHandler
import de.bitb.astroskop.ui.base.MVPFragment
import de.bitb.astroskop.ui.main.profiles.details.add.CreateConstellationActivity

import android.app.Activity.RESULT_OK
import de.bitb.astroskop.model.enums.Zodiac
import de.bitb.astroskop.ui.main.profiles.details.ProfileActivity.Companion.KEY_PROFILE_UUID
import de.bitb.astroskop.ui.main.profiles.details.add.CreateConstellationActivity.Companion.REQUEST_CREATE_CONSTELLATION

class ProfileDetailFragment : MVPFragment<IProfileDetailView, ProfileDetailPresenter>(), IProfileDetailView, IBind {

    @BindView(R.id.fragment_profile_detail_aries_list)
    var ariesList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_taurus_list)
    var taurusList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_gemini_list)
    var geminiList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_cancer_list)
    var cancerList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_leo_list)
    var leoList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_virgo_list)
    var virgoList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_libra_list)
    var libraList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_scorpio_list)
    var scorpioList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_sagittarius_list)
    var sagittariusList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_capricorn_list)
    var capricornList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_aquarius_list)
    var aquariusList: ViewGroup? = null
    @BindView(R.id.fragment_profile_detail_pisces_list)
    var piscesList: ViewGroup? = null

    override var actionbarCallback: ActionbarHandler.ActionbarCallback = object : ActionbarHandler.ActionbarCallback() {
            override val actionbarButton1Icon: Int
                get() = android.R.drawable.ic_menu_add

            override fun onActionbarButton1Clicked(): Boolean {
                CreateConstellationActivity.startActivity(activity!!)
                return true
            }
        }

    override val layoutId: Int
        get() = R.layout.fragment_profile_detail

    override fun createPresenter(application: BaseApplication): ProfileDetailPresenter {
        return ProfileDetailPresenter(application, this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter!!.onCreate(arguments!!.getString(KEY_PROFILE_UUID))
    }

    override fun initProfile(profile: Profile) {
        val constellations = profile.constellations
        ariesList!!.removeAllViews()
        taurusList!!.removeAllViews()
        geminiList!!.removeAllViews()
        cancerList!!.removeAllViews()
        leoList!!.removeAllViews()
        virgoList!!.removeAllViews()
        libraList!!.removeAllViews()
        scorpioList!!.removeAllViews()
        sagittariusList!!.removeAllViews()
        capricornList!!.removeAllViews()
        aquariusList!!.removeAllViews()
        piscesList!!.removeAllViews()
        for (constellation in constellations!!) {
            var viewGroup: ViewGroup? = null
            when (constellation.zodiac) {
                Zodiac.ARIES -> viewGroup = ariesList
                Zodiac.TAURUS -> viewGroup = taurusList
                Zodiac.GEMINI -> viewGroup = geminiList
                Zodiac.CANCER -> viewGroup = cancerList
                Zodiac.LEO -> viewGroup = leoList
                Zodiac.VIRGO -> viewGroup = virgoList
                Zodiac.LIBRA -> viewGroup = libraList
                Zodiac.SCORPIO -> viewGroup = scorpioList
                Zodiac.SAGITTARIUS -> viewGroup = sagittariusList
                Zodiac.CAPRICORN -> viewGroup = capricornList
                Zodiac.AQUARIUS -> viewGroup = aquariusList
                Zodiac.PISCES -> viewGroup = piscesList
            }
            val isHouse = constellation.isHouse
            val view = LayoutInflater.from(context).inflate(R.layout.profile_detail_constellation_list_item, null)
            var tv = view.findViewById<TextView>(R.id.profile_detail_list_item_label)
            tv.text = if (isHouse) "House:" else "Ruler: "
            tv = view.findViewById(R.id.profile_detail_list_item_text)
            tv.text = if (isHouse) constellation.house!!.name else constellation.ruler!!.name

            viewGroup!!.addView(view)
        }
    }

    override fun refreshView() {

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == RESULT_OK && requestCode == REQUEST_CREATE_CONSTELLATION) {
            presenter!!.addConstellation(data!!.getSerializableExtra(CreateConstellationActivity.KEY_CONSTELLATION) as Constellation)
        }
    }

    companion object {

        fun createInstance(uuid: String): ProfileDetailFragment {
            val frag = ProfileDetailFragment()
            val args = Bundle()
            args.putString(KEY_PROFILE_UUID, uuid)
            frag.arguments = args
            return frag
        }
    }
}
